library/blob.c

```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
-			eblob_dump_id(key->id), size, eblob_dump_dctl_flags(flags));
-
 	/* Sanity */
 	if (b == NULL || key == NULL) {
 		err = -EINVAL;
-		goto err_out_exit;
+		return err;
 	}
 
+	FORMATTED(HANDY_TIMER_SCOPE, ("eblob.%u.disk.write.prepare", b->cfg.stat_id));
+	EBLOB_WARNX(b->cfg.log, EBLOB_LOG_DEBUG,
+			"key: %s, size: %" PRIu64 ", flags: %s",
+			eblob_dump_id(key->id), size, eblob_dump_dctl_flags(flags));
+
 	/*
 	 * For eblob_write_prepare() this can fail with -E2BIG if we try to overwrite
 	 * record without footer by record with footer.
 	 */
 	pthread_mutex_lock(&b->lock);
 	defrag_generation = b->defrag_generation;
-
 	err = eblob_fill_write_control_from_ram(b, key, &wc, 1, &old);
-	pthread_mutex_unlock(&b->lock);
-	if (err && err != -ENOENT && err != -E2BIG)
+	if (err && err != -ENOENT && err != -E2BIG) {
+		pthread_mutex_unlock(&b->lock);
 		goto err_out_exit;
+	}
 
-	if (err == 0 && (wc.total_size >= eblob_calculate_size(b, key, 0, size))) {
-		uint64_t new_flags;
-
-		/*
-		 * We've found a key which will be overwritten,
-		 * make sure it has valid flags.
-		 *
-		 * We overwrite flags to what user has provided
-		 * dropping all existing on-disk flags since
-		 * given key will be fully overwritten and
-		 * we do not care about its old content anymore.
-		 *
-		 * The same logic is performed in @eblob_write_prepare_disk(),
-		 * but it also allocates new space.
-		 */
-		new_flags = eblob_validate_ctl_flags(b, flags);
-		new_flags |= BLOB_DISK_CTL_UNCOMMITTED;
-
-		if (wc.flags != new_flags) {
-			if (!(wc.flags & BLOB_DISK_CTL_UNCOMMITTED)) {
-				// update stats if the found key is committed, because now it becomes uncommitted.
-				eblob_stat_inc(wc.bctl->stat, EBLOB_LST_RECORDS_UNCOMMITTED);
-				eblob_stat_add(wc.bctl->stat, EBLOB_LST_UNCOMMITTED_SIZE,
-				               wc.total_size + sizeof(struct eblob_disk_control));
-
-				eblob_stat_inc(b->stat_summary, EBLOB_LST_RECORDS_UNCOMMITTED);
-				eblob_stat_add(b->stat_summary, EBLOB_LST_UNCOMMITTED_SIZE,
-				               wc.total_size + sizeof(struct eblob_disk_control));
-			}
-
-			wc.flags = new_flags;
-
-			err = eblob_commit_disk(b, key, &wc, 0);
-			if (err)
-				goto err_out_cleanup_wc;
-
-			err = eblob_commit_ram(b, key, &wc);
-			if (err)
-				goto err_out_cleanup_wc;
-		}
-
-		eblob_stat_inc(b->stat, EBLOB_GST_PREPARE_REUSED);
-		goto err_out_cleanup_wc;
+	// Note: were we have 3 cases:
+	// 1) We have not found a key. => hold backend lock, reserve space in the new base
+	// 2) We've found a key in existing bases.
+	//   2.a) Corresponding base has no enough space. => hold backend lock, reserve in the new base
+	//   2.b) Corresponding base has closed for writing (binlog enabled) => release backend lock,
+	//                                                                      reserve in the new base
+	if (err == 0 && (wc.total_size >= eblob_calculate_size(b, key, 0, size)) &&
+	    !eblob_binlog_enabled(&wc.bctl->binlog)) {
+		pthread_mutex_unlock(&b->lock);
```

Comment:
Shouldn't it be unlocked after `eblob_commit_disk`?

Reply:
It seem's that bctl critness equal 1 is enough.

Reply:
ok